subsys/net/ip/net_core.c

```diff
@@ -411,6 +411,56 @@ int net_try_send_data(struct net_pkt *pkt, k_timeout_t timeout)
 		goto err;
 	}
 
+
+#if defined(CONFIG_NET_IPV4)
+	if (net_pkt_family(pkt) == AF_INET &&
+	net_ipv4_is_addr_mcast((const struct in_addr *)&NET_IPV4_HDR(pkt)->dst)) {
+			struct net_context *ctx = net_pkt_context(pkt);
+
+		if (ctx && ctx->multicast_loop) {
+			struct net_pkt *clone = net_pkt_clone(pkt, K_NO_WAIT);
+
+			if (clone) {
+				NET_DBG("Looping back IPv4 multicast packet %p", clone);
+	
+				net_pkt_set_iface(clone, net_pkt_iface(pkt));
+	
+				if (net_recv_data(net_pkt_iface(clone), clone) < 0) {
+					NET_ERR("Failed to process cloned IPv4 packet");
+					net_pkt_unref(clone);
+				}
+			} 
+			else {
+					NET_ERR("Failed to clone IPv4 multicast packet");
+			}
+		}
+	}
+#endif
+
+#if defined(CONFIG_NET_IPV6)
+	if (net_pkt_family(pkt) == AF_INET6 &&
+	net_ipv6_is_addr_mcast((const struct in6_addr *)&NET_IPV6_HDR(pkt)->dst)) {
```

Comment:
IPv6 is controlled by a separate socket option, `IPV6_MULTICAST_LOOP`. @sayooj-aerlync was looking at that too.